email domain validation

misterpekert opened the pull request:
- Use google API to check if user email is valid

closes #29

edwh commented on components/EmailValidator.vue:
```diff
@@ -120,15 +122,30 @@ export default {
+        const domain = value.substring(value.indexOf('@') + 1)
+        const url = new URL('https://dns.google/resolve')
+        url.search = new URLSearchParams({ name: domain }).toString()
+        const googleResponse = await fetch(url).then((response) => response)
```
I'm slightly worried here about whether there are ad blockers which would cause this fetch never to complete.  If that happened, we'd never validate the address.  

Please can we add a timeout, say 10 seconds?  Then if the fetch doesn't complete we should just assume that the email is valid.

misterpekert replied:
I've added the new way of cancelling request using this API https://caniuse.com/mdn-api_abortsignal_timeout_static, but I'm not sure if that's okay in terms of browser support. If not, let me know and I'll rewrite with default setTimeout and AbortController to cancel the request)

edwh commented on components/EmailValidator.vue:
```diff
@@ -120,15 +122,30 @@ export default {
-        const validate = await this.$refs.email?.validate()
-        if (validate) {
-          this.$emit('update:valid', validate.valid)
+        const meta = this.$refs.form?.getMeta()
```
I'm a bit worried by `form?`.  If that ref doesn't exist at this point, wouldn't we just quietly fail and never emit?

I don't think it should happen - but if it shouldn't happen, we don't need the `?`.

misterpekert replied:
Yeah, `?` is extra here. It confuses me when coming from TS, 'cause there a ref can always be undefined and you always need to provide either `?` or `!` when you 100% know this item is in DOM at given moment.

edwh reviewed (changes requested):
You are probably right about the domain typo being the issue - though that must mean it wasn't tested fully.

However I think there is more that we can do to make sure that we definitely emit.

misterpekert commented:
@edwh you can check now, I've updated the PR.

misterpekert commented:
@edwh I've added a couple of new things to that PR:

- Fixed issue with email validation status being ignored by LoginModal (and other places as well, I think)
- Added domain cache (simple JS Map) so we don't query Google for the same domain
- Fixed some minor bugs related to validation to make the validation process more stable and predictable.

Would you mind giving that another try? I've checked places with emails I'm aware of, but I might have missed some scenarios that are not obvious to find :)

edwh commented on components/LoginModal.vue:
```diff
@@ -472,46 +472,50 @@ export default {
-                  .catch((err) => {
-                    console.error('Failed to save credentials', err)
+        if (this.emailError || this.passwordError) {
+          this.nativeLoginError = 'Please fill out the form.'
```
Is this the right error message?  We're inside `if (this.email && this.password)` - so we don't get this message in the case where the form hasn't been filled in - we just get the red borders:

![image](https://github.com/Freegle/iznik-nuxt3/assets/56696/14e295d1-13d1-425e-b2fa-5f945d74461c)

Should we just lose the `if (this.email && this.password` check?

misterpekert replied:
Is there any specific reason to send an API request with an invalid email/password? I've just added this check to avoid unnecessary calls to the API.

edwh replied:
There's no reason to send an invalid API call.  But when I tested the code, if you didn't fill in any forms and just hit the log in button then you get the red borders with no error message.  We need to show the error message.

misterpekert replied:
@edwh it seems that it's the same on production 
<img width="1250" alt="image" src="https://github.com/Freegle/iznik-nuxt3/assets/11361743/ff904c65-5a6c-4c30-b1cf-d3b5d614fb83">
I'll prepare a fix

misterpekert replied:
@edwh Done. https://github.com/Freegle/iznik-nuxt3/pull/50/commits/5e1c719c08fd33718e872acd8174c5bb20826368

edwh commented on components/EmailValidator.vue:
```diff
@@ -6,18 +6,20 @@
-        :placeholder="'Email address ' + (required ? '' : '(Optional)')"
-      />
-      <ErrorMessage name="email" class="text-danger font-weight-bold" />
+      <VeeForm ref="form" as="">
```
Are there any issues with nested `VeeForm`?  In `CommunityEventModal` we have `VForm` which is wrapping this.

I'm assuming `VForm` and `VeeForm` are the same - if so then we should use just `VeeForm` as it's clearer.

misterpekert replied:
We need Form wrapper from vee-validate because otherwise, it won't show <error-message> component. We can either implement it on our own and show/hide depending on validation result or rely on vee-validate, but we need <Form> wrapper.

I'll rename VForm to VeeFrom in other places as well.

edwh replied:
Does it introduce an actual `form` element into the DOM?  If it does then some uses of this component will result in a nested `form` element, which isn't valid HTML.

misterpekert replied:
No, it doesn't. I think having as="" makes it renderless

edwh replied:
Ok, great.  That's fine then.

edwh commented on components/EmailValidator.vue:
```diff
@@ -35,13 +37,15 @@
 import { ref } from '#imports'
 import { EMAIL_REGEX } from '~/constants'
 
+const domainValidationCache = new Map()
```
Feels overkill, since we are not likely to validate domains more than once before a page refresh.  But since it's there, leave it in.

misterpekert replied:
@edwh Well, vee-validate doesn't cache validation results, so that ends up querying Google every 'blur'/'change' event, like 2-3 requests per each "regex-valid" domain. I've thought having a small cache would help :)
